### PR TITLE
Added base support + sensors for IBM AMM

### LINKED ISSUE #3590: Add support for IBM AMM
Discovery run: http://pastebin.com/NB3cZ4Uf
Poller run: http://pastebin.com/vrQAimAf
SNMP walk: http://pastebin.com/rzvPCYnn
MIBs: https://drive.google.com/open?id=0B8tCbMiTmqN_aDBRcW1ScjhJTzQ

This is an IBM AMM- a management device for a blade chassis

If you could add support for it I'd be incredibly grateful.

Thanks,
burn

### PR DESCRIPTION
**Pull request by laf**
Fix #3590 

More can be added, the MIB is loaded with info but this is a starting point.

### CODE REVIEW

File: includes/discovery/sensors/fanspeeds/ibm-amm.inc.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by the
+ * Free Software Foundation, either version 3 of the License, or (at your
+ * option) any later version.  Please see LICENSE.txt at the top level of
+ * the source code distribution for details.
+ */
+
+if ($device['os'] == 'ibm-amm') {
+
+    $oids = array('blower1speedRPM', 'blower1speedRPM', 'blower1speedRPM', 'blower1speedRPM');
```
Comment (geordish) [inline includes/discovery/sensors/fanspeeds/ibm-amm.inc.php]: Should this not be blower1SpeedRPM, blower2speedRPM...?

File: includes/discovery/sensors/states/ibm-amm.inc.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+        if (!empty($state)) {
+
+            $state_name = $oid;
+            $state_index_id = create_state_index($state_name);
+            
+            if ($state_index_id) {
+
+                $states = array(
+                    array($state_index_id,'uknown',0,1,3) ,
```
Comment (murrant) [inline includes/discovery/sensors/states/ibm-amm.inc.php]: Typo s/uknown/unknown/
Maybe capitalize the first letter.

File: includes/discovery/sensors/fanspeeds/ibm-amm.inc.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    d_echo($oids."\n");
+    if (!empty($oids)) {
+
+        echo 'BLADE-MIB';
+        foreach ($oids as $data) {
+
+            if (!empty($data)) {
+                $value = snmp_get($device, $oid.'.0', '-OsqnU', 'BLADE-MIB');
```
Comment (geordish) [inline includes/discovery/sensors/fanspeeds/ibm-amm.inc.php]: After this, I think it should be exploding on " " into $oid and $value.

File: includes/discovery/sensors/fanspeeds/ibm-amm.inc.php [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+                if (is_numeric($value)) {
+                    $descr = $data;
+                    discover_sensor($valid['sensor'], 'fanspeed', $device, $data, 0, 'snmp', $descr, 1, 1, null, null, null, null, $value);
```
Comment (geordish) [inline includes/discovery/sensors/fanspeeds/ibm-amm.inc.php]: $data should be either $oid from above, or 'BLADE-MIB::' . $data

File: includes/discovery/sensors/fanspeeds/ibm-amm.inc.php [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    d_echo($oids."\n");
+    if (!empty($oids)) {
+
+        echo 'BLADE-MIB';
+        foreach ($oids as $data) {
+
+            if (!empty($data)) {
+                $value = snmp_get($device, $oid.'.0', '-OsqnU', 'BLADE-MIB');
+                if (is_numeric($value)) {
+                    $descr = $data;
```
Comment (geordish) [inline includes/discovery/sensors/fanspeeds/ibm-amm.inc.php]: This seems like a fairly useless assignment?

**Comment by rbax82** [comment]
Something's up, I have no idea what though.
http://puu.sh/pm1eg/99f52aae35.png

**Comment by laf** [comment]
@rbax82 please try again, you will need to update the branch of mine on your install:

```bash
cd /opt/librenms
git remote update laf
git checkout issue-3590
git pull laf issue-3590
```

File: includes/discovery/sensors/states/ibm-amm.inc.php [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+if ($device['os'] == 'ibm-amm') {
+
+    $index = 1;
+    $oids = array('blower1State','blower2State','blower3State','blower4State');
+
+    foreach ($oids as $oid) {
+
+        $state = snmp_get($device, $oid.'.0', '-OsqnU', 'BLADE-MIB');
+
+        if (!empty($state)) {
```
Comment (murrant) [inline includes/discovery/sensors/states/ibm-amm.inc.php]: Won't discover states that are 0, yet you define one below, is this intended?

File: includes/discovery/sensors/states/ibm-amm.inc.php [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    }//end foreach
+
+    $index = 1;
+    $oids = array('blower1ControllerState','blower2ControllerState','blower3ControllerState','blower4ControllerState');
+
+    foreach ($oids as $oid) {
+
+        $state = snmp_get($device, $oid.'.0', '-OsqnU', 'BLADE-MIB');
+
+        if (!empty($state)) {
```
Comment (murrant) [inline includes/discovery/sensors/states/ibm-amm.inc.php]: Won't discover sensors where $state == 0, but 0 is defined below as operational.

**Comment by rbax82** [comment]
http://puu.sh/pmETf/f1b1b1d460.png
Progress. =)

File: includes/discovery/sensors/fanspeeds/ibm-amm.inc.php [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+    $oids = array('blower1speedRPM', 'blower2speedRPM', 'blower3speedRPM', 'blower4speedRPM');
+    d_echo($oids."\n");
+    if (!empty($oids)) {
+
+        echo 'BLADE-MIB';
+        foreach ($oids as $data) {
+
+            if (!empty($data)) {
+                $value = snmp_get($device, $data.'.0', '-OsqnU', 'BLADE-MIB');
```
Comment (murrant) [inline includes/discovery/sensors/fanspeeds/ibm-amm.inc.php]: Add `v` to snmp_get options to only fetch the value.

File: includes/discovery/sensors/fanspeeds/ibm-amm.inc.php [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+                if (is_numeric($value)) {
```
Comment (murrant) [inline includes/discovery/sensors/fanspeeds/ibm-amm.inc.php]: `if (is_numeric(trim($value, '"'))) {`  pesky surrounding quotes.

**Comment by murrant** [comment]
@laf I don't know how to make the traffic graph in the upper right go away.  If you can fix that and merge my last pr, this is good to merge upstream.

image

**Comment by laf** [comment]
If that is you using the snmpsim then isn't it because if-mib isn't responding / working? If it's definitely not supported in the device we can remove the traffic graphs from the mini display.

**Comment by rbax82** [comment]
http://puu.sh/przyP/d33af006fa.png

**Comment by rbax82** [comment]
14:38:09 < lnms-bot> ./poller.php -r -f -m sensors  -h HOSTNAME -d
http://pastebin.com/1XJJNZqD

**Comment by laf** [comment]
bump

**Comment by murrant** [comment]
@rbax82 If you tested an earlier patch for this, you will need to remove the specific entries (or all) from the tables: `sensors`, `sensors_to_state_indexes`, `state_indexes`, `state_translations`

File: includes/discovery/sensors/states/ibm-amm.inc.php [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+            $state_index_id = create_state_index($state_name);
+            
+            if ($state_index_id) {
+
+                $states = array(
+                    array($state_index_id,'unknown',0,1,3) ,
+                    array($state_index_id,'good',1,2,0) ,
+                    array($state_index_id,'warning',1,3,1) ,
+                    array($state_index_id,'bad',1,4,2) ,
```
Comment (murrant) [inline includes/discovery/sensors/states/ibm-amm.inc.php]: Blower state indexes are off by one, should be 0,1,2,3 instead of 1,2,3,4

**Comment by laf** [comment]
Updated - thanks for the spot murrant.

**Comment by Rosiak** [comment]
👍

File: includes/discovery/os/ibm-amm.inc.php [hunk 11]
```diff
@@ -0,0 +1,5 @@
+<?php
+
+if (stristr($sysDescr, 'BladeCenter Advanced Management Module')) {
```
Comment (murrant) [inline includes/discovery/os/ibm-amm.inc.php]: One last review and I noticed this is not contained in a `if(!$os)` block

Comment (laf) [inline includes/discovery/os/ibm-amm.inc.php]: Done. Thanks.

**Pull request merged**